bam_addrprg.c

```diff
@@ -0,0 +1,471 @@
+        if (!strcmp(key,"ID")) {
+            char* retval = strdup(token);
+            free(line);
+            return retval;
+        }
+        token = strsep(&next,"\t");
+    }
+    free(line);
+    return NULL;
+}
+
+// Confirms the existance of an RG line with a given ID in a bam header
+static bool confirm_rg( const bam_hdr_t *hdr, const char* rgid )
+{
+    assert( hdr != NULL && rgid != NULL );
+
+    char *ptr, *start;
+    bool found = false;
+    start = ptr = strndup(hdr->text, hdr->l_text);
+    while (ptr != NULL && *ptr != '\0' && found == false ) {
+        size_t end = 0;
+        char* line = get_rg_line(ptr, &end);
+        if (line == NULL) break; // No more @RG
+        char* id;
+        if (((id = get_rg_id(line)) != NULL) && !strcmp(id, rgid)) {
+            found = true;
+        }
+        free(id);
+        free(line);
+        ptr += end;
+    }
+    free(start);
+    return found;
+}
+
+static char* get_first_rgid( const bam_hdr_t *hdr )
+{
+    assert( hdr != NULL );
+    char *ptr, *start;
+    char* found = NULL;
+    start = ptr = strndup(hdr->text, hdr->l_text);
+    while (ptr != NULL && *ptr != '\0' && found == NULL ) {
+        size_t end = 0;
+        char* line = get_rg_line(ptr, &end);
+        if ( line ) {
+            found = get_rg_id(line);
+        } else break;
+        free(line);
+        ptr += end;
+    }
+    free(start);
+    return found;
+}
+
+static void usage(FILE *fp)
+{
+    fprintf(fp,
+            "Usage: samtools addreplacerg [options] [-r <@RG line> | -R <existing id>] <input.bam> <output.bam>\n"
+            "\n"
+            "Options:\n"
+            "  -r STRING @RG line text\n"
+            "  -R STRING ID of @RG line in existing header to use\n"
+            "  -m MODE   Set the mode of operation from one of overwrite_all, orphan_only [overwrite_all]\n"
+            "  -l INT    Set compression level, from 0 (uncompressed) to 9 (best)\n");
+    sam_global_opt_help(fp, "..O..");
+}
+
+static bool parse_args(int argc, char** argv, parsed_opts_t** opts)
+{
+    *opts = NULL;
+    int n;
+    char modeout[12], *fmtout = NULL, *rg_line = NULL;
+    strcpy(modeout, "w");
+    int level = -1;
+
+    parsed_opts_t* retval = calloc(1, sizeof(parsed_opts_t));
+    if (! retval ) {
+        fprintf(stderr, "[%s] Out of memory allocating parsed_opts_t\n", __func__);
+        return true;
+    }
+    // Set defaults
+    retval->mode = overwrite_all;
+    sam_global_args_init(&retval->ga);
+    static const struct option lopts[] = {
+        SAM_OPT_GLOBAL_OPTIONS(0, 0, 'O', 0, 0),
+        { NULL, 0, NULL, 0 }
+    };
+
+    while ((n = getopt_long(argc, argv, "r:R:m:O:l:", lopts, NULL)) >= 0) {
+        switch (n) {
+            case 'r':
+                rg_line = optarg;
+                break;
+            case 'R':
+                retval->rg_id = strdup(optarg);
+                break;
+            case 'm': {
+                if (!strcmp(optarg, "overwrite_all")) {
```

Comment:
This one's going in that style document @dkj wants :smile:: [please don't](http://c-faq.com/style/strcmp.html), please consider writing `if (strcmp(...) == 0)` so we can see the `==`.